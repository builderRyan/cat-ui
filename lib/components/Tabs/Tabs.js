"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var react_1 = require("react");
var styled_1 = require("./styled");
var Tabs = (function (_super) {
    __extends(Tabs, _super);
    function Tabs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.toRender = function () {
            var children = _this.props.children;
            var navs = [];
            var contents = [];
            React.Children.forEach(children, function (child) {
                var _a = child.props, tab = _a.tab, id = _a.id;
                navs.push({ id: id, tab: tab });
                contents.push(child);
            });
            return { navs: navs, contents: contents };
        };
        _this.onTabClick = function (id) {
            var onChange = _this.props.onChange;
            onChange && onChange(id);
        };
        _this.toRenderNav = function (navs) {
            var activeId = _this.props.activeId;
            return navs.map(function (element) {
                var classes = activeId === element.id ? 'active' : '';
                return (React.createElement(styled_1.TabsNav, { key: element.id, onClick: function () { _this.onTabClick(element.id); }, className: classes }, element.tab));
            });
        };
        _this.toRenderContent = function (contents) {
            var activeId = _this.props.activeId;
            return contents.map(function (element, index) {
                var id = element.props.id;
                var classes = activeId === id ? 'active' : '';
                return (React.createElement(styled_1.TabsContent, { key: index, className: classes }, element));
            });
        };
        return _this;
    }
    Tabs.prototype.render = function () {
        var _a = this.props, className = _a.className, style = _a.style;
        var _b = this.toRender(), navs = _b.navs, contents = _b.contents;
        return (React.createElement(styled_1.StyledTabs, { className: className, style: style },
            React.createElement(styled_1.TabsNavBox, null, this.toRenderNav(navs)),
            React.createElement(styled_1.TabsContentBox, null, this.toRenderContent(contents))));
    };
    return Tabs;
}(react_1.Component));
exports.default = Tabs;
//# sourceMappingURL=Tabs.js.map