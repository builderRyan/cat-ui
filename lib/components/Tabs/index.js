"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Tabs_1 = require("./Tabs");
var TabPanel_1 = require("./TabPanel");
Tabs_1.default.TabPanel = TabPanel_1.default;
exports.default = Tabs_1.default;
//# sourceMappingURL=index.js.map