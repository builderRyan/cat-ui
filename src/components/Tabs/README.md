# Tabs 标签页

## Example

```jsx
<Tabs
  activeId={activeId}
  onChange={onTabChange1}>
  <TabPanel
    id='a'
    tab='选项A'>
    <div>选项A的内容11111</div>
  </TabPanel>
  <TabPanel
    id='b'
    tab='选项B'>
    <div>选项B的内容2222222</div>
  </TabPanel>
  <TabPanel
    id='c'
    tab='选项C'>
    <div>选项C的内容333333</div>
  </TabPanel>
</Tabs>
```

## API

|   属性    |             类型              | 默认值 |      描述      |
|:---------:|:-----------------------------:|:------:|:--------------:|
| className |           `string`            |   -    |      类名      |
|   style   |           `object`            |   -    |    行内样式    |
| activeId  |       `string | number`       |   -    | 激活的 tab id  |
| onChange  | `(id: string | number) => {}` |   -    | tab 切换的回调 |

