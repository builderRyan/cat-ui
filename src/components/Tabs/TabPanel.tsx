import * as React from 'react'
import { Component } from 'react'
// import { StyledTabPanel } from './styled'

export interface TabPanelProps {
  /** 类名 */
  className?: string;
  /** 样式 */
  style?: object;
  /** 该TabPanel的id */
  id: string | number;
  /** 该TabPanel所对应的tab标签的名字	 */
  tab: string;
}

/**
 * TabPanel
 */
class TabPanel extends Component<TabPanelProps, any> {

  render() {
    return (
      <div>
        TabPanel
      </div>
    )
  }
}

export default TabPanel
