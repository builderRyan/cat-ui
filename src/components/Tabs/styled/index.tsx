import styled from 'styled-components'
import { primaryBtnColor } from '../../../styles/theme'

const StyledTabs = styled.div`

`;

const TabsNavBox = styled.div`
  border-bottom: 1px solid #eeeeee;
  margin-bottom: 20px;
`;

const TabsNav = styled.div`
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  height: 100%;
  padding: 12px 0;
  text-decoration: none;
  cursor: pointer;
  transition: color .3s cubic-bezier(.645, .045, .355, 1);
  font-size: 18px;
  color: #4a4a4a;

  &.active {
    font-weight: bold;
    &::before {
      position: absolute;
      left: 0;
      bottom: 0;
      content: "";
      width: 40px;
      height: 3px;
      background-color: #ff4444;
    }
  }

  & + & {
    margin-left: 40px;
  }
`;

const TabsContentBox = styled.div`
`;

const TabsContent = styled.div`
  display: none;

  &.active {
    display: block;
  }
`;

export { StyledTabs, TabsNavBox, TabsNav, TabsContentBox, TabsContent }

