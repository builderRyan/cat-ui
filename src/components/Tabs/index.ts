import Tabs from './Tabs'
import TabsPanel from './TabPanel'

Tabs.TabsPanel = TabsPanel

export default Tabs
