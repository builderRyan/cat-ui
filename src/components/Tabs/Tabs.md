Example

```jsx
import TabPanel from './TabPanel';

class Example extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      activeId1: 'a',
      activeId2: 'd'
    }
    this.onTabChange1 = this.onTabChange1.bind(this)
    this.onTabChange2 = this.onTabChange2.bind(this)
  }

  onTabChange1 (id) {
    this.setState({
      activeId1: id
    });
  }

  onTabChange2 (id) {
    this.setState({
      activeId2: id
    });
  }

  render () {
    const { activeId1, activeId2 } = this.state

    return (<div>
      <h3>基础Tabs</h3>
      <Tabs 
        activeId={activeId1}
        onChange={this.onTabChange1}>
        <TabPanel 
          id='a'
          tab='选项A'>
          <div>选项A的内容11111</div>
        </TabPanel>
        <TabPanel 
          id='b'
          tab='选项B'>
          <div>选项B的内容2222222</div>
        </TabPanel>
        <TabPanel 
          id='c'
          tab='选项C'>
          <div>选项C的内容333333</div>
        </TabPanel>
      </Tabs>

      <h3>TabPanel的tab属性为dom && TabPanel的children为空</h3>

      <Tabs 
        activeId={activeId2}
        onChange={this.onTabChange2}>
        <TabPanel 
          id='d'
          tab={(<a href='/#table'>选项table</a>)}>
        </TabPanel>
        <TabPanel 
          id='e'
          tab={(<a href='/#slider'>选项slider</a>)}>
        </TabPanel>
        <TabPanel 
          id='f'
          tab={(<span href='/#tooltip'>选项<span>tooltip</span></span>)}>
        </TabPanel>
      </Tabs>

      <style>
        {`

        `}
      </style>
    </div>
    )
  }
}

<Example />
```

