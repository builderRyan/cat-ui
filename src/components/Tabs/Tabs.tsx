import * as React from 'react'
import { Component } from 'react'
import { StyledTabs, TabsNavBox, TabsNav, TabsContentBox, TabsContent } from './styled'
import TabPanel from './TabPanel'

export interface TabsProps {
  /** 类名 */
  className?: string;
  /** 样式 */
  style?: object;
  /** 激活的tab-key */
  activeId?: string | number;
  /** 选中的tab改变时 */
  onChange?: (id: string | number) => {};
}

/**
 * Tabs
 */
class Tabs extends Component<TabsProps, any> {
  static TabsPanel = TabPanel;

  toRender = () => {
    const { children } = this.props
    let navs: Object[] = []
    let contents: Object[] = []
    React.Children.forEach(children, (child: any) => {
      const { props: { tab, id } } = child
      navs.push({id: id, tab: tab})
      contents.push(child)
    })
    return { navs, contents }
  }

  onTabClick = (id: string | number) => {
    const { onChange } = this.props
    onChange && onChange(id)
  }

  toRenderNav = (navs: Object[]) => {
    const { activeId } = this.props
    return navs.map((element: any) => {
      let classes = activeId === element.id ? 'active': ''
      return (
        <TabsNav
          key={element.id}
          onClick={() => {this.onTabClick(element.id)}}
          className={classes}>
          {element.tab}
        </TabsNav>
      )
    })
  }

  toRenderContent = (contents: any) => {
    const { activeId } = this.props
    return contents.map((element: any, index: number) => {
      const { id } = element.props
      let classes = activeId === id ? 'active': ''
      return (
        <TabsContent
          key={index}
          className={classes}>
          {element}
        </TabsContent>
      )
    })
  }

  render() {
    let { className, style } = this.props
    let { navs, contents } = this.toRender()

    return (
      <StyledTabs
        className={className}
        style={style}>
        <TabsNavBox>
          {this.toRenderNav(navs)}
        </TabsNavBox>
        <TabsContentBox>
          {this.toRenderContent(contents)}
        </TabsContentBox>
      </StyledTabs>
    )
  }
}

export default Tabs

