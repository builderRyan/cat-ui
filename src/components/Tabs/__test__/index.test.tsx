import * as React from 'react';
import * as Enzyme from 'enzyme';
import { render } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import Tabs from '../index';

Enzyme.configure({ adapter: new Adapter() });

describe('Tabs', () => {
  test('renders correctly', () => {
    const wrapper = render(<Tabs activeId='1'>Test</Tabs>);
    expect(wrapper).toMatchSnapshot();

    const wrapper2 = render(
      <Tabs activeId='1' tabBarExtraContent={'tabBarExtraContent'}>
        Test
      </Tabs>
    );
    expect(wrapper2).toMatchSnapshot();
  });
});
