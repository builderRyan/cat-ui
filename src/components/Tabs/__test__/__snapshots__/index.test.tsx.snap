// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "cat-tabs",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "cat-tabs__nav-wrapper",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "cat-tabs__navs",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "cat-tabs__content-wrapper",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "cat-tabs__content-wrapper",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "cat-tabs__nav-wrapper",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "cat-tabs__navs",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Tabs renders correctly 2`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "cat-tabs",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "cat-tabs__nav-wrapper",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "cat-tabs__navs",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "cat-tabs__extra-content",
              },
              "children": Array [
                Object {
                  "data": "tabBarExtraContent",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "cat-tabs__extra-content",
            },
            "children": Array [
              Object {
                "data": "tabBarExtraContent",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "cat-tabs__navs",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "cat-tabs__content-wrapper",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "cat-tabs__content-wrapper",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "cat-tabs__nav-wrapper",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "cat-tabs__navs",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "cat-tabs__extra-content",
                },
                "children": Array [
                  Object {
                    "data": "tabBarExtraContent",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            Object {
              "attribs": Object {
                "class": "cat-tabs__extra-content",
              },
              "children": Array [
                Object {
                  "data": "tabBarExtraContent",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "cat-tabs__navs",
                },
                "children": Array [],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
